M43 Sunfury Lasgun MK1

Surrealaser opened the pull request:
-Adds the M43 Sunfury Lasgun

-Long range focused battle rifle: high accuracy, no recoil or scatter while wielded (horrible accuracy and scatter while one-handed).
Comes standard with UGL.
50 rounds.
Long wield time (1 second).
Overcharge and Regular mode
Regular: 3.33 shots / sec, 25 burn damage, 30 penetration per round. 83.33 DPS
Overcharge: 1 shot / sec, 55 burn damage, 60 penetration, 2 ammo consumed per round. 55 DPS

As a basis of comparison to the pulse rifle:
Single: 2.5 shots / sec. 40 brute damage, 10 penetration per round, 100 DPS.
Burst: 3.75 shots / sec. 40 brute damage, 10 penetration per round, 150 DPS.

-New ColMarTech Lasgun Field Charger vendor. Recharges restocked lasgun batteries with an internal battery that recharges off the power grid; can be supply dropped like a crate.

Surrealaser committed:
M43 Sunfury Lasgun MK1

-Adds the M43 Sunfury Lasgun

-Long range focused battle rifle: high accuracy, no recoil or scatter while wielded (horrible accuracy and scatter while one-handed).
Comes standard with miniscope.
50 rounds.
Long wield time (1 second).
Overcharge and Regular mode
Regular: 3.33 shots / sec, 30 burn damage, 30 penetration per round. 100 DPS
Overcharge: 1 shot / sec, 80 burn damage, 60 penetration, 2 ammo consumed per round. 80 DPS

As a basis of comparison to the pulse rifle:
Single: 2.5 shots / sec. 40 brute damage, 10 penetration per round, 100 DPS.
Burst: 3.75 shots / sec. 40 brute damage, 10 penetration per round, 150 DPS.

-New ColMarTech Lasgun Field Charger vendor. Recharges restocked lasgun batteries with an internal battery that recharges off the power grid; can be supply dropped like a crate.

factoryman942 commented:
on-ground sprite doesn't exist

factoryman942 commented:
also light laser is oriented perpendicular to how it should be

factoryman942 commented:
the overcharge laser is hilariously strong against walls; six of them takes down a rwall, and another two opens the girder to be shot through

Supermichael777 commented:
Its too strong single shot, its accuracy and high burst make it substantially above parity with other standard equipment.  It is about as effective as incendary sniper, but your handing one to every boot

Surrealaser commented:
All of these issues are related to the hasty 4th rebase I had to do to get this out in time.

This is why we need to coordinate when we update the master.

Surrealaser committed:
Nerf for the lasgun

Overcharge damage reduced to 55.

Miniscope replaced with UGL.

Rest is minor formatting fixes.

Surrealaser committed:
Nerfed Overcharge damage to 40

Surrealaser commented:
Alright, nerfed the overcharge damage people were complaining about.

Fixed the sprite issues.

Rohesie commented on code/game/machinery/vending/marine_vending.dm:
```diff
@@ -232,7 +238,62 @@
+	req_access = null
+	req_access_txt = "0"
+	req_one_access = null
+	req_one_access_txt = "9;2;21"
```
Favor `req_one_access` over `req_one_access_txt`.

Rohesie commented on code/game/machinery/vending/marine_vending.dm:
```diff
@@ -232,7 +238,62 @@
+	update_icon()
+
+/obj/machinery/vending/lasgun/update_icon()
+	if(machine_current_charge <= 0)
```
Use `switch(machine_current_charge*100/machine_max_charge)`

Rohesie replied:
Add a `if(machine_max_charge)` first, though, to avoid a possible runtime error if you try to divide by zero.

Rohesie commented on code/game/machinery/vending/vending.dm:
```diff
@@ -543,12 +543,21 @@
-		sleep(delay_vending)
+		if(powered(power_channel))
+			use_power(vend_power_usage)	//actuators and stuff
+			if (icon_vend) flick(icon_vend,src) //Show the vending animation if needed
```
Same-line control statement.

Rohesie commented on code/game/machinery/vending/vending.dm:
```diff
@@ -562,9 +571,12 @@
 	if(istype(A, /obj/item))
 		var/obj/item/I = A
-		stock(I, user)
+		if(istype(I, /obj/item/ammo_magazine/lasgun) && istype(src, /obj/machinery/vending/lasgun))
```
Better to redefine this proc for your specific vending machine instead.

Surrealaser committed:
New test damage values

Standard charge damage lowered to 25/30

Overcharge damage lowered to 55/60

Rohesie commented on code/modules/projectiles/updated_projectiles/gun_system.dm:
```diff
@@ -217,16 +222,17 @@
-			return
-			
+	if(user.get_inactive_hand())//Drop item held in the off-hand if occupied.
+		var/obj/item/offhand = user.get_inactive_hand()
```
Not to call the same proc twice you can define the var first:
`var/obj/item/offhand`
And then run the control statement: `if(offhand = user.get_inactive_hand())`
It will check the proc, give the value to `offhand` and proceed if not null or false.

Rohesie commented on code/modules/projectiles/updated_projectiles/gun_system.dm:
```diff
@@ -466,10 +475,26 @@ load_into_chamber(), reload_into_chamber(), and clear_jam() do all of the heavy
 	//The workhorse of the bullet procs.
  	//If we have a round chambered and no active attachable, we're good to go.
 	if(in_chamber && !active_attachable)
+		if(overcharge_check) //Check to see if we have the proper ammo in chamber to match the overcharge fire mode
```
This is a hack to implement energy weapon code on a clearly ballistics system. Use an energy-based system, it will be cleaner and simpler.

Surrealaser replied:
Let me know which functions/function changes you want me to port over to energy.dm; I think some like the cocking ones and ammo counter feedback are fine.

Surrealaser committed:
Switchify icon update for lascharger vendor

Surrealaser committed:
Formatting Changes

Surrealaser committed:
Migrated lasgun specific functions to local procs

Surrealaser committed:
Porting over more procs

Surrealaser committed:
Reformatting etc

Surrealaser committed:
Migrated procs from gun_system to energy

Surrealaser committed:
Minor formatting changes

Surrealaser committed:
Conflict resolving sprite sheets

Rohesie commented on code/game/machinery/vending/marine_vending.dm:
```diff
@@ -232,7 +238,73 @@
+		stock(W, user, TRUE)
+		//to_chat(user, "<span class='warning'>DEBUG: Lasgun vendor restocked with [W].</span>")
+		return TRUE
+	. = ..()
```
`return ..()` if at the end of the proc.

Rohesie commented on code/game/machinery/vending/marine_vending.dm:
```diff
@@ -232,7 +238,73 @@
+/obj/machinery/vending/lasgun/attackby(obj/item/W, mob/user)
+	if(istype(W, /obj/item/ammo_magazine/lasgun))
+		stock(W, user, TRUE)
+		//to_chat(user, "<span class='warning'>DEBUG: Lasgun vendor restocked with [W].</span>")
```
Clear.

Rohesie commented on code/game/machinery/vending/marine_vending.dm:
```diff
@@ -232,7 +238,73 @@
+/obj/machinery/vending/lasgun/MouseDrop_T(var/atom/movable/A, mob/user)
+	..()
 
+	if(istype(A, /obj/item))
```
This check is superfluous.

Rohesie commented on code/game/machinery/vending/marine_vending.dm:
```diff
@@ -232,7 +238,73 @@
 
+	if(istype(A, /obj/item))
+		var/obj/item/I = A
+		if(istype(I, /obj/item/ammo_magazine/lasgun) && istype(src, /obj/machinery/vending/lasgun))
```
The second check is superfluous. This is a redefined proc.

Rohesie commented on code/game/machinery/vending/vending.dm:
```diff
@@ -543,12 +543,23 @@
+			sleep(delay_vending)
+		else
+			return
 	if(ispath(R.product_path,/obj/item/weapon/gun)) . = new R.product_path(get_turf(src),1)
```
`return` instead of `.` at the end of the proc.

Rohesie commented on code/game/machinery/vending/vending.dm:
```diff
@@ -543,12 +543,23 @@
+		else
+			return
 	if(ispath(R.product_path,/obj/item/weapon/gun)) . = new R.product_path(get_turf(src),1)
 	else . = new R.product_path(get_turf(src))
```
`return` instead of `.` at the end of the proc.

Rohesie commented on code/game/machinery/vending/vending.dm:
```diff
@@ -543,12 +543,23 @@
-		use_power(vend_power_usage)	//actuators and stuff
-		if (icon_vend) flick(icon_vend,src) //Show the vending animation if needed
-		sleep(delay_vending)
+		if(powered(power_channel))
```
This is an interesting system you are adding for vendors. Tie it with an actual battery and it's golden.

Surrealaser replied:
Why do I need an actual battery instead of a more efficient variable?

ninjanomnom replied:
It's cleaner and more extendable. In game it also makes a lot more sense if the power supply is shared with other similar devices.

Rohesie commented on code/game/machinery/vending/vending.dm:
```diff
@@ -581,7 +592,22 @@
 				var/obj/item/ammo_magazine/A = item_to_stock
-				if(A.current_rounds < A.max_rounds)
+				//to_chat(user, "<span class='warning'>DEBUG: Magazine Name: [A]. Recharge?: [recharge]. Current Charge: [machine_current_charge].</span>")
+				if(istype(A, /obj/item/ammo_magazine/lasgun) && recharge)
```
If you are using `istype()` chances are you should be doing a more specific proc instead.

Rohesie replied:
No, I meant something like /obj/machinery/vending/your_vending_machine/MouseDrop_T

Rohesie commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -245,7 +245,9 @@
 					dat += "<B>Launch Pad Status:</b> "
-					var/obj/structure/closet/crate/C = locate() in current_squad.drop_pad.loc
+					var/obj/C = locate() in current_squad.drop_pad.loc //This thing should ALWAYS exist.
+					if(!istype(C,/obj/machinery/vending) && !istype(C,/obj/structure/closet/crate)) //Can only send vendors and crates
```
This is a big hack. Sending down a vending machine?

Surrealaser replied:
What's the issue?

Surrealaser replied:
I mean if you want to restrict it to the lasgun recharger, that's fine; this exception is largely here for that, though medvends on supply drop would be pretty nice.

Rohesie commented on code/modules/projectiles/updated_projectiles/ammunition.dm:
```diff
@@ -19,6 +19,7 @@ They're all essentially identical when it comes to getting the job done.
 	throw_speed = 2
 	throw_range = 6
 	var/default_ammo = /datum/ammo/bullet
+	var/overcharge_ammo = /datum/ammo/bullet //Generally used for energy weapons
```
This is a hack and it should be using the energy system.

Surrealaser replied:
What energy system? You mean the one from TG that I'm definitely not porting over?

ninjanomnom replied:
If you're not willing to port or rewrite large systems yourself don't even bother adding these kinds of features which if done improperly introduce large amounts of code debt.

Rohesie commented on code/modules/projectiles/updated_projectiles/gun_attachables.dm:
```diff
@@ -684,6 +684,19 @@ obj/item/attachable/attack_hand(var/mob/user as mob)
+	name = "\improper M43 Sunfury lasgun stock"
+	desc = "The standard stock for the M43 Sunfury lasgun."
+	slot = "stock"
+	wield_delay_mod = null
```
Add wield delay if you are giving a melee mod.

Surrealaser replied:
It has a 1 second wield delay; this is higher even than the HPR after its wield delay nerf.

Rohesie commented on code/modules/projectiles/updated_projectiles/gun_system.dm:
```diff
@@ -60,7 +60,11 @@
+	var/extra_delay		= 0						//When burst-firing, this number is extra time before the weapon can fire again. Depends on number of rounds fired.
+
+	//Energy Weapons
+	var/ammo_per_shot	= 1						//How much ammo consumed per shot; normally 1.
```
This should be dealt by energy systems.

Rohesie commented on code/modules/projectiles/updated_projectiles/gun_system.dm:
```diff
@@ -291,7 +296,8 @@
 		ammo = ammo_list[/datum/ammo/bullet] //Looks like we're defaulting it.
 	else
-		ammo = ammo_list[magazine.default_ammo]
+		ammo = ammo_list[overcharge? magazine.overcharge_ammo : magazine.default_ammo]
```
Same here, this shouldn't be processed by all guns.

Rohesie commented on code/modules/projectiles/updated_projectiles/gun_system.dm:
```diff
@@ -528,7 +536,7 @@ and you're good to go.
 		cdel(projectile_to_fire) //Getting rid of it. Attachables only use ammo after the cycle is over.
 		if(refund)
-			active_attachable.current_rounds++ //Refund the bullet.
+			active_attachable.current_rounds += ammo_per_shot //Refund the bullet.
```
This would be so much more elegantly dealt with by a battery.

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,235 @@
+						/obj/item/attachable/verticalgrip,
+						/obj/item/attachable/angledgrip,
+						/obj/item/attachable/lasersight,
+						/obj/item/attachable/gyro,
```
What's with all these attachments?

Surrealaser replied:
I'm not sure why these would be inappropriate.

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,235 @@
+
+//Toggles Overcharge mode. Overcharge mode significantly increases damage and AP in exchange for doubled ammo usage and increased fire delay.
+/obj/item/weapon/gun/energy/lasgun/proc/toggle_chargemode(mob/user)
+	if(overcharge == 0)
```
`FALSE`

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,235 @@
+		damage_falloff_mult = config.low_damage_falloff_mult
+		fire_sound = 'sound/weapons/Laser3.ogg'
+		to_chat(user, "\icon[src] You [overcharge? "<B>disable</b>" : "<B>enable</b>" ] [src]'s overcharge mode.")
+		overcharge = 1
```
`TRUE`

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,235 @@
+		//While overcharge is active, double ammo consumption, and
+		playsound(user, 'sound/weapons/emitter.ogg', 15, 1)
+		ammo_per_shot = OVERCHARGE_AMMO_COST
+		fire_delay = config.med_fire_delay * 2 // 1 shot per second fire rate
```
Higher, if it's going to be perfect-accuracy long-range 55 damage 55 AP.

Surrealaser replied:
How much higher? Will you not be happy until the DPS is lower than a slug?

Rohesie replied:
For sure! A slug has much lower accuracy and can't be shot through a scope. Compare it to a sniper gun and bullet and make it weaker, much weaker.

Surrealaser replied:
And it also knocks down, slows and disables, and is accurate enough to impose all of these effects consistently at the range it can do so.

You don't use slugs for the DPS, you use them for the spike damage and the debuffs.

Damage is the _only_ thing the overcharge has.

Rohesie replied:
Besides sniper range, full accuracy and a ton of AP. Unless this is a variant for the sniper spec, tweak it.

Surrealaser replied:
AP that's only relevant vs a handful of xenos, and in practice unless you get a scope, 2 tiles more effective range.

At best the damage could use a minor tweak downward, but I'm not doubling the fire delay or doing whatever other crazy thing you had in mind.

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,235 @@
+			item_state = "m43_0_w"
+		else
+			item_state = "m43_0"
+	else if(current_mag.current_rounds > round(current_mag.max_rounds*0.75))
```
Use a `switch`

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,235 @@
+
+/obj/item/weapon/gun/energy/lasgun/replace_ammo(mob/user = null, var/obj/item/ammo_magazine/magazine)
+	if(!magazine.default_ammo)
+		to_chat(user, "Something went horribly wrong. Ahelp the following: ERROR CODE A1: null ammo while reloading.")
```
Use a message to admins instead of telling the user to AHelp.

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,235 @@
+		//to_chat(user, "DEBUG: REPLACE AMMO. Ammo: [ammo]")
+
+
+/obj/item/weapon/gun/energy/lasgun/load_into_chamber(mob/user, overcharge_check = FALSE)
```
The need for this should be using an energy-based system obvious.

Surrealaser commented on code/game/machinery/machinery.dm:
```diff
@@ -104,6 +81,8 @@ Class Procs:
 		//2 = run auto, use active
 	var/idle_power_usage = 0
 	var/active_power_usage = 0
+	var/machine_current_charge = 0 //Does it have an integrated battery? Normally 10k if so.
```
How about no and we stick with a more efficient variable instead of pointlessly gluing a battery to the vendor?

spookydonut replied:
```suggestion
	var/machine_current_charge = 0 //Does it have an integrated, unremovable capacitor? Normally 10k if so.
```

Surrealaser committed:
Update vending.dm

Surrealaser committed:
Update marine_vending.dm

Surrealaser committed:
Update machinery.dm

Surrealaser committed:
Update gun_attachables.dm

Surrealaser committed:
Update energy.dm

Surrealaser committed:
Update energy.dm

Rohesie commented on code/game/machinery/vending/vending.dm:
```diff
@@ -556,8 +556,8 @@
 			return
-	if(ispath(R.product_path,/obj/item/weapon/gun)) . = new R.product_path(get_turf(src),1)
-	else . = new R.product_path(get_turf(src))
+	if(ispath(R.product_path,/obj/item/weapon/gun)) return = new R.product_path(get_turf(src),1)
```
Same line control statement.

Rohesie commented on code/game/machinery/vending/vending.dm:
```diff
@@ -556,8 +556,8 @@
-	if(ispath(R.product_path,/obj/item/weapon/gun)) . = new R.product_path(get_turf(src),1)
-	else . = new R.product_path(get_turf(src))
+	if(ispath(R.product_path,/obj/item/weapon/gun)) return = new R.product_path(get_turf(src),1)
+	else return = new R.product_path(get_turf(src))
```
Same line control statement.

Borisvanmemes commented:
Is this PR dead in the water?
It looks a tad bit dead in the water.

Surrealaser commented:
@Borisvanmemes It's waiting on code ports, though it will definitely be added at some point

spookydonut reviewed (changes requested):
Overall this is fine, it's either do this or bolt some crazy energy bolt system onto the projectile/hit/damage system which would be fucking awful.

spookydonut commented on code/modules/cm_marines/overwatch.dm:
```diff
@@ -245,7 +245,9 @@
 					dat += "<B>Launch Pad Status:</b> "
-					var/obj/structure/closet/crate/C = locate() in current_squad.drop_pad.loc
+					var/obj/C = locate() in current_squad.drop_pad.loc //This thing should ALWAYS exist.
+					if(!istype(C,/obj/machinery/vending) && !istype(C,/obj/structure/closet/crate)) //Can only send vendors and crates
```
```suggestion
					if(!istype(C,/obj/machinery/vending/lasgun) && !istype(C,/obj/structure/closet/crate)) //Can only send vendors and crates
```

spookydonut commented on code/game/machinery/vending/vending.dm:
```diff
@@ -726,3 +741,18 @@
 		if (WIRE_SHOOTINV)
 			src.shoot_inventory = !src.shoot_inventory
+
+/obj/machinery/vending/proc/recharge_lasguncell(/obj/item/ammo_magazine/lasgun/A, mob/user)
```
```suggestion
/obj/machinery/vending/lasgun/proc/recharge_lasguncell(/obj/item/ammo_magazine/lasgun/A, mob/user)
```

spookydonut commented on code/game/machinery/vending/vending.dm:
```diff
@@ -581,7 +592,10 @@
 				var/obj/item/ammo_magazine/A = item_to_stock
-				if(A.current_rounds < A.max_rounds)
+				//to_chat(user, "<span class='warning'>DEBUG: Magazine Name: [A]. Recharge?: [recharge]. Current Charge: [machine_current_charge].</span>")
+				if(istype(A, /obj/item/ammo_magazine/lasgun) && recharge)
```
Move this to a `/obj/machinery/vending/lasgun/stock()`

Surrealaser committed:
Seal removal

Surrealaser committed:
Merge remote-tracking branch 'ColonialMarines-Mirror/master' into LasgunUpdate3

Surrealaser committed:
Update Icons

Surrealaser committed:
Minor Fixes

spookydonut committed:
Update code/game/machinery/machinery.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

spookydonut committed:
Update code/modules/cm_marines/overwatch.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Rohesie commented on code/game/machinery/machinery.dm:
```diff
@@ -191,14 +170,41 @@ Class Procs:
-		return 0
-	if(src.use_power == 1)
-		use_power(idle_power_usage,power_channel, 1)
+		if(src.use_power == TRUE && (machine_current_charge > idle_power_usage)) //Does it have an integrated battery/reserve power to tap into?
```
```suggestion
		if(use_power && (machine_current_charge > idle_power_usage)) //Does it have an integrated battery/reserve power to tap into?
```

Rohesie commented on code/game/machinery/machinery.dm:
```diff
@@ -191,14 +170,41 @@ Class Procs:
+		else
+			return FALSE
+
+	if(src.use_power == TRUE)
```
```suggestion
	if(use_power)
```

Rohesie commented on code/game/machinery/vending/marine_vending.dm:
```diff
@@ -240,6 +245,72 @@
+	return ..()
+
+/obj/machinery/vending/lasgun/examine(mob/user)
+	..()
```
```suggestion
	. = ..()
```

Rohesie replied:
This is still up.

Rohesie commented on code/game/machinery/vending/marine_vending.dm:
```diff
@@ -240,6 +245,72 @@
+
+
+/obj/machinery/vending/lasgun/MouseDrop_T(var/atom/movable/A, mob/user)
+	..()
```
```suggestion
	. = ..()
```

Rohesie replied:
This is still up.

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,236 @@
+	current_mag.current_rounds = max(0,current_mag.current_rounds - amount)
+	update_icon()
+	current_mag.update_icon()
+	..()
```
```suggestion
	return ..()
```

Rohesie commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,236 @@
+// use power from a cell
+/obj/item/ammo_magazine/lasgun/proc/use(var/amount)
+
+	return 1
```
```suggestion
	return TRUE
```

Rohesie commented on code/modules/projectiles/updated_projectiles/magazines/energy.dm:
```diff
@@ -0,0 +1,70 @@
+	var/amount = round(max_rounds * rand(2,severity) * 0.1)
+	current_rounds = max(0,current_rounds - amount)
+	update_icon()
+	..()
```
```suggestion
	return ..()
```

Rohesie committed:
Update code/game/machinery/machinery.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Rohesie committed:
Update code/game/machinery/machinery.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Rohesie committed:
Update code/game/machinery/vending/marine_vending.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Rohesie committed:
Update code/game/machinery/vending/marine_vending.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Rohesie committed:
Update code/modules/projectiles/updated_projectiles/guns/energy.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Rohesie committed:
Update code/modules/projectiles/updated_projectiles/guns/energy.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Rohesie committed:
Update code/modules/projectiles/updated_projectiles/magazines/energy.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Surrealaser committed:
Update Icons and Requested Changes

Surrealaser committed:
Lasgun supply pack

Surrealaser committed:
Merge remote-tracking branch 'ColonialMarines-Mirror/master' into LasgunUpdate3

Surrealaser committed:
More annoying icon conflict resolution

Surrealaser committed:
Merge remote-tracking branch 'ColonialMarines-Mirror/master' into LasgunUpdate3

Surrealaser committed:
Revert removal of whitespace from documentation

Surrealaser committed:
Resolving residual issues from prior merge

ninjanomnom commented:
Do it properly or not at all, the feature isn't worth the amount of code debt it introduces by duplicating logic and adding in new systems. Refactor base code to make extension natural.

ninjanomnom commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,236 @@
+	if(current_mag && current_mag.current_rounds > 0)
+		if(current_mag.current_rounds < ammo_per_shot && overcharge)
+			if(istype(src, /obj/item/weapon/gun/energy/lasgun))
+				var/obj/item/weapon/gun/energy/lasgun/L = src
```
...Why are you checking that src is a lasgun inside a redefined proc for the lasgun

ninjanomnom commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,236 @@
+//EMPs will fuck with remaining charge
+/obj/item/weapon/gun/energy/lasgun/emp_act(severity)
+	var/amount = round(current_mag.max_rounds * rand(2,severity) * 0.1)
+	if(current_mag.current_rounds < amount)	return 0
```
Don't use single line control statements

ninjanomnom commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,236 @@
+	current_mag.current_rounds = max(0,current_mag.current_rounds - amount)
+	update_icon()
+	current_mag.update_icon()
+	return ..()
```
Is there some reason this needs to come last in this proc? Prefer to use `. = ..()` first instead when you can

ninjanomnom commented on code/modules/projectiles/updated_projectiles/magazines/energy.dm:
```diff
@@ -0,0 +1,70 @@
+//Energy weapons
+
+/*
+/obj/item/weapon_cell
```
Please don't add in code just to comment it out

ninjanomnom commented on code/modules/projectiles/updated_projectiles/magazines/energy.dm:
```diff
@@ -0,0 +1,70 @@
+/obj/item/ammo_magazine/lasgun/M43/update_icon()
+	switch(current_rounds / max(1,max_rounds))
+		if(0)
+			icon_state = base_ammo_icon + "_0"
```
```
var/remaining = CEILING((current_rounds / max(1,max_rounds)) * 100, 25)
icon_state = "[base_ammo_icon]_[remaining]"
```

ninjanomnom commented on code/modules/projectiles/updated_projectiles/magazines/energy.dm:
```diff
@@ -0,0 +1,70 @@
+			icon_state = base_ammo_icon + "_50"
+		if(0.01 to 0.25)
+			icon_state = base_ammo_icon + "_25"
+	//to_chat(world, "<span class='warning'>DEBUG: Lasgun Magazine Icon Update. Icon State: [icon_state] Current Rounds: [current_rounds]</span>")
```
As before

Surrealaser committed:
Merge remote-tracking branch 'ColonialMarines-Mirror/master' into LasgunUpdate3

Rohesie committed:
Update code/game/machinery/vending/marine_vending.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Rohesie committed:
Update code/game/machinery/vending/marine_vending.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Surrealaser commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,236 @@
+			//to_chat(user, "<span class='warning'>DEBUG: Load Into Chamber. Overcharge: [overcharge] Chamber: [in_chamber] Reg_Ammo: [reg_ammo] Over_Ammo: [over_ammo] </span>")
+			if(overcharge && in_chamber.name == reg_ammo.name)
+				//to_chat(user, "<span class='warning'>DEBUG: Chamber ammo replaced with overcharge. Chamber: [in_chamber] </span>")
+				in_chamber = null //clean the chamber of the erroneous round
```
If the chambered ammo type matches the mode setting, then we don't clear the chamber.

ninjanomnom committed:
Update code/modules/projectiles/updated_projectiles/guns/energy.dm

Co-Authored-By: Surrealaser <42553659+Surrealaser@users.noreply.github.com>

Surrealaser committed:
Requested changes

Surrealaser commented on code/modules/projectiles/updated_projectiles/guns/energy.dm:
```diff
@@ -371,3 +371,236 @@
+			item_state = "m43_0"
+	else
+		switch(current_mag.current_rounds / max(current_mag.max_rounds, 1))
+			if(0.76 to 1)
```
This is done.

Surrealaser committed:
Merge remote-tracking branch 'ColonialMarines-Mirror/master' into LasgunUpdate3

Surrealaser committed:
Tweaked highcap laspacks to be more in line with other extended mags

Per title

Surrealaser committed:
Corrected minor storage, sound and aesthetic issues

Surrealaser committed:
Belt too

Surrealaser committed:
Minor bug and sprite fixes.
